apps/aecore/src/aec_id.erl

```diff
@@ -0,0 +1,104 @@
+%%% -*- erlang-indent-level:4; indent-tabs-mode: nil -*-
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2018, Aeternity Anstalt
+%%% @doc
+%%% ADT for identifiers
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(aec_id).
+
+-export([ create/2
+        , specialize/1
+        , specialize/2
+        , specialize_type/1
+        ]).
+
+%% For aec_serialization
+-export([ encode/1
+        , decode/1
+        ]).
+
+-record(id, { tag
+            , val
+            }).
+
+-type tag() :: 'account' | 'oracle' | 'name'
+             | 'commitment' | 'contract' | 'channel'.
+-type val() :: <<_:256>>.
+-opaque(id() :: #id{}).
+
+-export_type([ id/0
+             , tag/0
+             , val/0
+             ]).
+
+-define(PUB_SIZE, 32).
+-define(TAG_SIZE, 1).
+-define(SERIALIZED_SIZE, 33). %% ?TAG_SIZE + ?PUB_SIZE
+
+-define(IS_TAG(___TAG___), ___TAG___ =:= account;
+                           ___TAG___ =:= oracle;
+                           ___TAG___ =:= name;
+                           ___TAG___ =:= commitment;
+                           ___TAG___ =:= contract;
+                           ___TAG___ =:= channel
+       ).
+-define(IS_VAL(___VAL___), byte_size(___VAL___) =:= 32).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec create(tag(), val()) -> id().
+create(Tag, Val) when ?IS_TAG(Tag), ?IS_VAL(Val) ->
+    #id{ tag = Tag
+       , val = Val};
+create(Tag, Val) when ?IS_VAL(Val) ->
+    error({illegal_tag, Tag});
+create(Tag, Val) when ?IS_TAG(Tag)->
+    error({illegal_val, Val}).
```

Comment:
What, if it is neither a tag nor a value?

Reply:
You get a lovely `function_clause` ;-)